chainstate/test-suite/src/tests/pos_processing_tests.rs

```diff
@@ -419,17 +484,451 @@ fn pos_invalid_pool_id(#[case] seed: Seed) {
+        stake_pool_outpoint,
+        &vrf_sk,
+        // no epoch is sealed yet so use initial randomness
+        initial_randomness,
+        pool_id,
+        1,
+    );
+    let reward_output = TxOutput::new(
+        OutputValue::Coin(Amount::from_atoms(1)),
+        OutputPurpose::ProduceBlockFromStake(Box::new(stake_pool_data.clone())),
+    );
     tf.make_block_builder()
-        .with_consensus_data(ConsensusData::PoS(pos_data))
+        .with_consensus_data(ConsensusData::PoS(Box::new(pos_data)))
+        .with_reward(vec![reward_output])
+        .build_and_process()
+        .unwrap();
+
+    // prepare and process block_c with ProduceBlockFromStake -> ProduceBlockFromStake kernel
+    let block_a_reward_outpoint = OutPoint::new(OutPointSourceId::BlockReward(block_a_id), 0);
+    let pos_data = create_pos_data(
+        &mut tf,
+        block_a_reward_outpoint,
+        &vrf_sk,
+        // no epoch is sealed yet so use initial randomness
+        initial_randomness,
+        pool_id,
+        1,
+    );
+    let reward_output = TxOutput::new(
+        OutputValue::Coin(Amount::from_atoms(1)),
+        OutputPurpose::ProduceBlockFromStake(Box::new(stake_pool_data.clone())),
+    );
+    let block_c_index = tf
+        .make_block_builder()
+        .with_consensus_data(ConsensusData::PoS(Box::new(pos_data)))
+        .with_reward(vec![reward_output])
+        .with_parent(block_a_id)
         .build_and_process()
         .unwrap()
         .unwrap();
+
+    // prepare and process block_d with ProduceBlockFromStake -> ProduceBlockFromStake kernel
+    let block_3_reward_outpoint = OutPoint::new(
+        OutPointSourceId::BlockReward((*block_c_index.block_id()).into()),
+        0,
+    );
+
+    // both sealed epoch and pre block randomness can be used
+    let sealed_epoch_randomness =
+        tf.storage.transaction_ro().unwrap().get_epoch_data(1).unwrap().unwrap();
+    let pos_data = create_pos_data(
+        &mut tf,
+        block_3_reward_outpoint,
+        &vrf_sk,
+        sealed_epoch_randomness.randomness().value(),
+        pool_id,
+        2,
+    );
+    let reward_output = TxOutput::new(
+        OutputValue::Coin(Amount::from_atoms(1)),
+        OutputPurpose::ProduceBlockFromStake(Box::new(stake_pool_data)),
+    );
+    tf.make_block_builder()
+        .with_consensus_data(ConsensusData::PoS(Box::new(pos_data)))
+        .with_reward(vec![reward_output])
+        .with_parent((*block_c_index.block_id()).into())
+        .build_and_process()
+        .unwrap();
+
+    let res_pool_balance =
+        PoSAccountingStorageRead::<TipStorageTag>::get_pool_balance(&tf.storage, pool_id)
+            .unwrap()
+            .unwrap();
+    let total_subsidy =
+        tf.chainstate.get_chain_config().block_subsidy_at_height(&BlockHeight::from(1)) * 3;
+    let initially_staked = Amount::from_atoms(1);
+    assert_eq!(
+        (total_subsidy.unwrap() + initially_staked).unwrap(),
+        res_pool_balance
+    );
 }
```

Comment:
Please also add a test for using bogus targets to have an easy difficulty.

Reply:
I'll do these tests when I add target calculation for PoS